### DISCUSSION TITLE
[Question] compose non-DB async operation in DBIO transaction

### DISCUSSION THREAD

**Discussion by sebarys** [slick/slick#2858] [Questions] [answered]
Hi,

I have question about how combine DB related DBIOAction and other async actions e.g. calling upstream service using http client.

I want implement following flow
- `DBIO1` - select from DB users 
- `Future1` - execute HTTP call to retrieve some data from upstream service
- `DBIO2` - update some of users retrieved in `DBIO1` based on data from `Future1`
- `Future2` - execute HTTP call to upstream service

To compose `DBIO1` and `DBIO2` in single transaction I can combine them in bigger DBIO and use `.transactionally` https://scala-slick.org/doc/stable/dbio.html#transactions-and-pinned-sessions-transactions-

My question is how I can include in this DBIO also Future based stuff. I saw `DBIOAction.from` method 
```
  /** Convert a `Future` to a [[DBIOAction]]. */
  def from[R](f: Future[R]): DBIOAction[R, NoStream, Effect] = FutureAction[R](f)
```
1 . Is it recommended approach to embed non-DB async operations in `Future` context inside DBIOAction that should be executed transactionally?
2. In my example if `Future2` that is converted to `DBIO` (using `DBIOAction.from`) will fail, it means that whole DB interaction will be rollback?
3. Is is possible to pass explicitly `ExecutionContext` for non-DB related actions? I don't want use DB `AsyncExecutor` for non-DB related actions such as HTTP interaction with upstream service.

Thanks in advance for help or recommendation how such flow should be implemented when using Slick library!

**Comment by nafg** [comment, marked as answer]
> Is it recommended approach to embed non-DB async operations in Future context inside DBIOAction that should be executed transactionally?

I don't see any other way, given the requirement of it being transactional. I mean you could write lower-level code but what would be the benefit? It's true that the more time elapses between steps in a transaction the more overlap there could be with other operations, but AFAIK that's just a part of life.

> In my example if Future2 that is converted to DBIO (using DBIOAction.from) will fail, it means that whole DB interaction will be rollback?

I believe so

> Is is possible to pass explicitly ExecutionContext for non-DB related actions? I don't want use DB AsyncExecutor for non-DB related actions such as HTTP interaction with upstream service.

Whenever you compose `DBIO`s, e.g. with `flatMap` or `>>`, just like when you compose `Future`s it takes an implicit `ExecutionContext`. So I guess it will use whatever `ExecutionContext` is in implicit scope there (or that you pass in explicitly).
